AppDB/appscale/datastore/fdb/transactions.py

```diff
@@ -0,0 +1,305 @@
+from google.appengine.datastore import entity_pb
+
+logger = logging.getLogger(__name__)
+
+
+class TransactionMetadata(object):
+  """
+  A TransactionMetadata directory handles the encoding and decoding details for
+  transaction metadata for a specific project.
+
+  The directory path looks like (<project-dir>, 'transactions').
+  Within this directory, keys are encoded as
+  <scatter-byte> + <txid> + <rpc-type (optional)> + <rpc-details (optional)>.
+
+  The <scatter-byte> is a single byte derived from the txid. Its purpose is to
+  spread writes more evenly across the cluster and minimize hotspots. This is
+  especially important for this index because each write is given a new, larger
+  <txid> value than the last.
+
+  The <txid> is an 8-byte integer that serves as a handle for the client to
+  identify a transaction. It also serves as a read versionstamp for FDB
+  transactions used within the datastore transaction.
+
+  The initial creation of the datastore transaction does not specify any RPC
+  details. The purpose of that KV is to verify that the datastore transaction
+  exists (and the garbage collector hasn't cleaned it up) before committing it.
+
+  The <rpc-type> is a single byte that indicates what kind of RPC is being
+  logged as having occurred inside the transaction.
+
+  The <rpc-details> encodes the necessary details in order for the datastore
+  to reconstruct the RPCs that occurreed during the transaction when it comes
+  time to commit the mutations.
+
+  # TODO: Go into more detail about how different RPC types are encoded.
+  """
+  DIR_NAME = u'transactions'
+
+  LOOKUPS = b'\x00'
+  QUERIES = b'\x01'
+  PUTS = b'\x02'
+  DELETES = b'\x03'
+
+  # The max number of bytes for each FDB value.
+  _CHUNK_SIZE = 10000
+
+  _ENTITY_LEN_SIZE = 3
+
+  def __init__(self, directory):
+    self.directory = directory
+
+  @property
+  def project_id(self):
+    return self.directory.get_path()[len(DS_ROOT)]
+
+  @classmethod
+  def directory_path(cls, project_id):
+    return project_id, cls.DIR_NAME
+
+  def encode_start_key(self, scatter_val, commit_versionstamp=None):
+    key = b''.join([self.directory.rawPrefix, six.int2byte(scatter_val),
+                    commit_versionstamp or b'\x00' * VERSIONSTAMP_SIZE])
+    if not commit_versionstamp:
+      key += encode_versionstamp_index(len(key) - VERSIONSTAMP_SIZE)
+
+    return key
+
+  def encode_lookups(self, txid, keys):
+    section_prefix = self._txid_prefix(txid) + self.LOOKUPS
+    return self._encode_chunks(section_prefix, self._encode_keys(keys))
+
+  def encode_query_key(self, txid, namespace, ancestor_path):
+    section_prefix = self._txid_prefix(txid) + self.QUERIES
+    encoded_ancestor = Text.encode(namespace) + Path.pack(ancestor_path[:2])
+    return section_prefix + encoded_ancestor
+
+  def encode_puts(self, txid, entities):
+    section_prefix = self._txid_prefix(txid) + self.PUTS
+    encoded_entities = [entity.Encode() for entity in entities]
```

Comment:
Unfortunately `.Encode()` does perform work to reserialize the entity.

It's true that the serialized entity is present in the `PutRequest` that the datastore server receives. However, extracting the serialized entity from the encoded `PutRequest` would require a specialized parser. At this time, I think it would be a premature optimization to implement that. That said, if re-serialization turns out to be a performance problem even after switching to the new protobuf library, we can definitely consider doing something like that.